### DISCUSSION TITLE
ImportError related to the cryptography library

### DISCUSSION THREAD

**Discussion by winter-warm-sun** [mitmproxy/mitmproxy#7107] [General]
#### Problem Description
This issue occurs when running a script agent in the background on a server and opening a proxy on the client using an IP and port number. After some time, frequent errors begin to appear the following day. The expected behavior is for the system to operate normally, but the actual behavior results in an ImportError related to the cryptography library.

#### Steps to Reproduce
1.The script agent executes in the background，Deploy on the server
2.Open the proxy on the client by IP+port number
3.Frequent errors will occur the next day
Tried through the command pip install cryptography -U update the corresponding dependencies，
the problem could be solved and executed normally at that time, but the same error would occur again the next day.

#### System Information
Mitmproxy: 10.4.2
Python:    3.10.9
OpenSSL:   OpenSSL 3.2.2 4 Jun 2024
Platform:  Linux-3.10.107-1-tlinux2_kvm_guest-0054-x86_64-with-glibc2.17

#### Relevant log output
Traceback (most recent call last):
File "/usr/local/bin/mitmdump", line 5, in
from mitmproxy.tools.main import mitmdump
File "/usr/local/lib/python3.10/site-packages/mitmproxy/tools/main.py", line 15, in
from mitmproxy import master
File "/usr/local/lib/python3.10/site-packages/mitmproxy/master.py", line 5, in
from .addons import termlog
File "/usr/local/lib/python3.10/site-packages/mitmproxy/addons/init.py", line 6, in
from mitmproxy.addons import clientplayback
File "/usr/local/lib/python3.10/site-packages/mitmproxy/addons/clientplayback.py", line 25, in
from mitmproxy.proxy import layers
File "/usr/local/lib/python3.10/site-packages/mitmproxy/proxy/layers/init.py", line 3, in
from .http import HttpLayer
File "/usr/local/lib/python3.10/site-packages/mitmproxy/proxy/layers/http/init.py", line 14, in
from ..quic import QuicStreamEvent
File "/usr/local/lib/python3.10/site-packages/mitmproxy/proxy/layers/quic.py", line 13, in
from aioquic.h3.connection import ErrorCode as H3ErrorCode
File "/usr/local/lib/python3.10/site-packages/aioquic/h3/connection.py", line 19, in
from aioquic.quic.connection import QuicConnection, stream_is_unidirectional
File "/usr/local/lib/python3.10/site-packages/aioquic/quic/connection.py", line 21, in
from .. import tls
File "/usr/local/lib/python3.10/site-packages/aioquic/tls.py", line 41, in
from cryptography.hazmat.primitives.asymmetric.types import (
ImportError: cannot import name 'CertificateIssuerPublicKeyTypes' from 'cryptography.hazmat.primitives.asymmetric.types' (/usr/local/lib/python3.10/site-packages/cryptography/hazmat/primitives/asymmetric/types.py)

**Comment by mhils** [comment]
What's your output of `pip freeze`?

**Reply by winter-warm-sun** [reply]
My output is as follows
(weak-network) root@VM-89-17-centos ~/weak-network-project>pip freeze
aiohappyeyeballs==2.3.5
aiohttp==3.10.3
aioice==0.9.0
aioquic==1.2.0
aiortc==1.9.0
aiosignal==1.3.1
aircv==1.4.6
anyio==3.6.2
appnope==0.1.3
arrow==1.2.3
asgiref==3.8.1
asn1==2.6.0
astor==0.8.1
asttokens==2.2.1
async-timeout==4.0.2
attrdict==2.0.1
attrs==22.1.0
av==12.3.0
Babel==2.14.0
backcall==0.2.0
bce-python-sdk==0.9.6
beautifulsoup4==4.12.3
blessed==1.19.1
blinker==1.6.2
bpylist2==4.0.1
Brotli==1.1.0
cached-property==1.5.2
cachetools==5.3.3
certifi==2022.12.7
cffi==1.15.1
chardet==3.0.4
charset-normalizer==2.1.1
click==8.1.3
cmd2==2.4.2
coloredlogs==15.0.1
construct==2.10.68
contourpy==1.2.1
cryptography==38.0.4
cssselect==1.2.0
cssutils==2.10.2
cycler==0.12.1
Cython==3.0.10
daemonize==2.5.0
decorator==5.1.1
Deprecated==1.2.13
Django==5.0.7
dnspython==2.2.1
email-validator==1.3.0
enum-compat==0.0.3
et-xmlfile==1.1.0
exceptiongroup==1.2.0
executing==1.2.0
facebook-wda==1.4.6
fastapi==0.88.0
fire==0.6.0
Flask==3.0.0
flask-babel==4.0.0
flatbuffers==24.3.25
fonttools==4.50.0
frozenlist==1.4.1
future==1.0.0
google-crc32c==1.5.0
gpxpy==1.5.0
grpcio==1.60.0
gunicorn==22.0.0
GytestReport @ file:///root/devops_environment/workspace/build_tmp/GytestReport_e-df698a3cc8ae4c6684365886b8f7fefb_data3931346445859043668/GytestReport-1.0.0.tar.gz#sha256=383005cb2d6ca0070a4970b0cccaf1cbba262b445660c1672dafbdce7596bd5c
h11==0.14.0
h2==4.1.0
hexdump==3.3
hpack==4.0.0
httpcore==0.16.2
httptools==0.5.0
httpx==0.23.1
humanfriendly==10.0
hyperframe==6.0.1
idna==2.7
ifaddr==0.2.0
imageio==2.31.1
imageio-ffmpeg==0.4.7
imgaug==0.4.0
importlib-metadata==6.8.0
iniconfig==2.0.0
inquirer==3.0.0
ipsw-parser==1.1.5
ipython==8.7.0
itsdangerous==2.1.2
jedi==0.18.2
Jinja2==3.1.2
kaitaistruct==0.10
kiwisolver==1.4.5
la_panic==0.4.9
lazy_loader==0.3
ldap3==2.9.1
lmdb==1.4.1
loguru==0.7.2
lxml==5.2.1
MarkupSafe==2.1.1
matplotlib==3.8.3
matplotlib-inline==0.1.6
mitmproxy==10.4.2
mitmproxy_rs==0.6.3
mpmath==1.3.0
msgpack==1.0.8
multidict==6.0.5
nest-asyncio==1.5.6
networkx==3.1
numpy==1.24.3
onnxruntime==1.16.3
opencv-contrib-python==4.6.0.66
opencv-python==4.7.0.72
opencv-python-headless==4.9.0.80
openpyxl==3.1.2
opt-einsum==3.3.0
orjson==3.8.3
packaging==22.0
paddleocr==2.7.3
paddlepaddle==2.6.1
pandas==2.2.1
parso==0.8.3
passlib==1.7.4
pdf2docx==0.5.8
pexpect==4.8.0
pickleshare==0.7.5
Pillow==9.3.0
pillow-heif==0.13.1
pluggy==1.3.0
plumbum==1.8.2
premailer==3.10.0
prompt-toolkit==3.0.36
protobuf==5.27.3
psutil==5.9.4
ptyprocess==0.7.0
publicsuffix2==2.20191221
pure-eval==0.2.2
py==1.11.0
pyasn1==0.6.0
pyasn1_modules==0.4.0
pyclipper==1.3.0.post5
pycparser==2.21
pycrashreport==0.0.9
pycryptodome==3.16.0
pydantic==1.10.2
pyee==11.1.0
Pygments==2.13.0
pygnuutils==0.0.6
pyimg4==0.7
pykdebugparser==1.2.4
pyliblzfse==0.4.1
pylibsrtp==0.10.0
pylsqpack==0.3.18
pylzss==0.3.1
pymobiledevice3==1.34.2
PyMuPDF==1.24.1
PyMuPDFb==1.24.1
pyOpenSSL==24.2.1
pyparsing==3.0.9
pyperclip==1.8.2
pytesseract==0.3.10
pytest==7.4.4
python-dateutil==2.8.2
python-docx==1.1.0
python-dotenv==0.21.0
python-editor==1.0.4
python-multipart==0.0.5
python_atom_sdk==1.2.9
python_scriptexecute @ file:///root/devops_environment/workspace/build_tmp/pythonscriptrun_e-e7128161aaec4bf79f13661dde22b109_data139608518884369546/python_scriptexecute-6.0.9.tar.gz#sha256=54a89f351d603a949065caae568fc84a16ef61685b53b925656362127a64497f
pytz==2023.3
pyusb==1.2.1
PyWavelets==1.4.1
PyYAML==6.0
rainbow-sdk==1.1.12
rapidfuzz==3.7.0
rapidocr-onnxruntime==1.3.16
rarfile==4.1
readchar==4.0.3
remotezip==0.12.1
requests==2.20.1
requests-toolbelt==0.9.1
retry==0.9.2
rfc3986==1.5.0
ruamel.yaml==0.18.6
ruamel.yaml.clib==0.2.8
scikit-image==0.22.0
scipy==1.10.1
service-identity==24.1.0
shapely==2.0.3
six==1.16.0
sniffio==1.3.0
sortedcontainers==2.4.0
soupsieve==2.5
sqlparse==0.5.1
stack-data==0.6.2
starlette==0.22.0
sympy==1.12
tabulate==0.9.0
termcolor==2.1.1
tifffile==2023.7.10
tomli==2.0.1
tornado==6.4.1
tqdm==4.64.1
traitlets==5.7.1
typing_extensions==4.4.0
tzdata==2023.3
ujson==5.6.0
urllib3==1.24.3
urwid==2.6.15
urwid-mitmproxy==2.1.2.1
uvicorn==0.20.0
uvloop==0.17.0
visualdl==2.5.3
watchfiles==0.18.1
wcwidth==0.2.5
websocket-client==1.4.2
websockets==10.4
Werkzeug==3.0.1
wrapt==1.14.1
wsproto==1.2.0
yarl==1.9.4
zeroconf==0.71.4
zipp==3.16.2
zstandard==0.23.0
(weak-network) root@VM-89-17-centos ~/weak-network-project>
